### ISSUE TITLE
[Feature Request] Dilation and Groups Arguments could be added to Conv2d

### ISSUE THREAD

**Issue by Aavache** [ml-explore/mlx#100]
# 🚀 What is the purpose of this issue?

According to the [documentation](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.nn.Conv2d.html), these are the parameters available for `Conv2d`.

* **in_channels** ([int](https://docs.python.org/3/library/functions.html#int)) – The number of input channels.
* **out_channels** ([int](https://docs.python.org/3/library/functions.html#int)) – The number of output channels.
* **kernel_size** ([int](https://docs.python.org/3/library/functions.html#int) or [tuple](https://docs.python.org/3/library/stdtypes.html#tuple)) – The size of the convolution filters.
* **stride** ([int](https://docs.python.org/3/library/functions.html#int) or [tuple](https://docs.python.org/3/library/stdtypes.html#tuple), optional) – The size of the stride when applying the filter. Default: 0.
* **padding** ([int](https://docs.python.org/3/library/functions.html#int) or [tuple](https://docs.python.org/3/library/stdtypes.html#tuple), optional) – How many positions to 0-pad the input with. Default: 0.
* **bias** ([bool](https://docs.python.org/3/library/functions.html#bool), optional) – If True add a learnable bias to the output. Default: True

However, the **dilation** and **groups** parameter as defined in the torch implementation of [Conv2d](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html), would be desirable to implement well-known architectures, e.g. ASPP. The following is the description of these two parameters according to torch:

- **dilation** controls the spacing between the kernel points; also known as the à trous algorithm. It is harder to describe, but this [link](https://github.com/vdumoulin/conv_arithmetic/blob/master/README.md) has a nice visualization of what dilation does.

- **groups** control the connections between inputs and outputs. in_channels and out_channels must both be divisible by groups. For example, at groups=1, all inputs are convolved to all outputs. At groups=2, the operation becomes equivalent to having two conv layers side by side, each seeing half the input channels and producing half the output channels, and both subsequently concatenated.

Thank you for your attention 👍

**Comment by dc-dc-dc** [comment]
@jagrit06 in [conv2d](https://github.com/ml-explore/mlx/blob/4fe2fa2a6404725f6cab743d74a1c71d85f45844/mlx/ops.cpp#L2722) op call, there is an assert to check if dilations is set to 1, but from a quick look through the implementation it seems to support dilations. I removed this check and messed with it a little comparing to pytorch and am seeing the same results with different values for dilations. Is there a specific case where dilations isn't supported?

**Comment by jagrit06** [comment]
Well at the moment the convolutions with dilations revert to a fairly naive and slow kernel on the GPU (though I think we do fine on the CPU)

That restriction was added to basically side step that performance hit

**Comment by dc-dc-dc** [comment]
Should the assert be removed and create a separate issue to improve Convolution speed on GPU when dilation is not 1? Creates more visibility to the core issue and opens the possibility of a contributor able to help out

**Comment by a1eaiactaest** [comment]
agree with @dc-dc-dc, this would be really helpful for implementations that **require** dilation

**Comment by Jimexist** [comment]
FWIW [this][1] PR now also depends on a `group` param present.

[1]: https://github.com/ml-explore/mlx-examples/pull/405

**Comment by awni** [comment]
Is anyone working on groups for convolution? E.g. @Stealeristaken ?

I am closing https://github.com/ml-explore/mlx/pull/637 as it's inactive but it might be useful as a starting point.

**Comment by Stealeristaken** [comment]
> Is anyone working on groups for convolution? E.g. @Stealeristaken ?
> 
> I am closing #637 as it's inactive but it might be useful as a starting point.

Been a bit busy these days. I tried adding groups to convs, but it didn't work very well. So, I closed that pull request. I'll try again in the next few days.

**Comment by Rifur13** [comment]
I'm eager to get this working as well. I can pick this up if no one is working on it.

**Comment by awni** [comment]
I don't think anyone is actively working on it! Would be great to have you work on it @Rifur13 !

**Comment by Stealeristaken** [comment]
> I'm eager to get this working as well. I can pick this up if no one is working on it.

feel free to take. gl!

**Comment by Maalvi14** [comment]
Issue can be closed. Both dilation (added in PR #766) and groups (added in PR #1129) are implemented and tested for Conv2d.
